### PR TITLE
Stub Device Plugin refactor: 600x speedup

### PR DESCRIPTION
**Pull request by RenaudWasTaken**
**What this PR does / why we need it**:

Real size of PR is +65 −40
Real impact of the PR:
- Current test time fixed 112s
- Test time after this PR: 0.206
- That's around 600x speedup

Found a few more things which were not closed correctly:
- Depends on #52616 for re-registration test
- panics if the gRPC server in stub returns with an error before the end of the test
- Closes the stream in ListAndWatch
- Stub Device Plugin now uses glog

**Which issue this PR fixes**: fixes #52561

**Special notes for your reviewer**:
- I tried to keep changes to a minimum.
- Please only check the last commit, comments on the first commits should be directed to #52616 
- @vishh @jiayingz @mindprince 

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by feiskyer** [comment]
> Replaced the mutex in manager.go by a sync.Map to fix data race in go test

sync.Map is introduced in go1.9, but kubernetes is still building with go1.8.3. I think this should wait a while until go1.8.x is not supported.

**Comment by RenaudWasTaken** [comment]
@jiayingz can you take a look ?
I've vastly improved speed and reliability of the tests in this new version.

Running the tests with `-count 30` now takes 30-32 seconds instead of > 10 minutes and none of them are flaky.

**Comment by heeeello** [comment]
unsubscribe


________________________________
发件人: k8s-ci-robot <notifications@github.com>
发送时间: 2017年10月19日 4:44
收件人: kubernetes/kubernetes
抄送: Subscribed
主题: Re: [kubernetes/kubernetes] Fix data race and go routine leak in Device Plugin tests (#52612)


@RenaudWasTaken<https://github.com/renaudwastaken>: The following tests failed, say /retest to rerun them all:

Test name       Commit  Details Rerun command
pull-kubernetes-verify  ac6b25f<https://github.com/kubernetes/kubernetes/commit/ac6b25f36ea1d356d10cea8106cb80fe1b47a54e>       link<https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/52612/pull-kubernetes-verify/52005/>      /test pull-kubernetes-verify
pull-kubernetes-node-e2e        ac6b25f<https://github.com/kubernetes/kubernetes/commit/ac6b25f36ea1d356d10cea8106cb80fe1b47a54e>       link<https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/52612/pull-kubernetes-node-e2e/51745/>    /test pull-kubernetes-node-e2e
pull-kubernetes-e2e-gce 405ebfa<https://github.com/kubernetes/kubernetes/commit/405ebfac29912abc7134df4ce8df9324ff9fdcd1>       link<https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/52612/pull-kubernetes-e2e-gce/249/>       /test pull-kubernetes-e2e-gce
pull-kubernetes-e2e-gce-gpu     405ebfa<https://github.com/kubernetes/kubernetes/commit/405ebfac29912abc7134df4ce8df9324ff9fdcd1>       link<https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/52612/pull-kubernetes-e2e-gce-gpu/10298/> /test pull-kubernetes-e2e-gce-gpu
pull-kubernetes-bazel-build     405ebfa<https://github.com/kubernetes/kubernetes/commit/405ebfac29912abc7134df4ce8df9324ff9fdcd1>       link<https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/52612/pull-kubernetes-bazel-build/10230/> /test pull-kubernetes-bazel-build

Full PR test history<https://k8s-gubernator.appspot.com/pr/52612>. Your PR dashboard<https://k8s-gubernator.appspot.com/pr/RenaudWasTaken>. Please help us cut down on flakes by linking to<https://github.com/kubernetes/community/blob/master/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests> an open issue<https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open> when you hit one in your PR.

Instructions for interacting with me using PR comments are available here<https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md>. If you have questions or suggestions related to my behavior, please file an issue against the kubernetes/test-infra<https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:> repository. I understand the commands that are listed here<https://github.com/kubernetes/test-infra/blob/master/commands.md>.

―

File: pkg/kubelet/cm/deviceplugin/device_plugin_stub.go [hunk 1]
```diff
@@ -143,7 +182,7 @@ func (m *Stub) Update(devs []*pluginapi.Device) {
 
 // Allocate does a mock allocation
 func (m *Stub) Allocate(ctx context.Context, r *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
-	log.Printf("Allocate, %+v", r)
+	glog.V(2).Infof("Allocate, %+v", r)
```
Comment (vikaschoudhary16) [inline pkg/kubelet/cm/deviceplugin/device_plugin_stub.go]: would it make sense to use a higher log level eg, 4 ?

File: pkg/kubelet/cm/deviceplugin/device_plugin_stub.go [hunk 2]
```diff
@@ -109,12 +147,13 @@ func (m *Stub) Register(kubeletEndpoint, resourceName string) error {
...
 		return err
 	}
+
 	return nil
 }
 
 // ListAndWatch lists devices and update that list according to the Update call
 func (m *Stub) ListAndWatch(e *pluginapi.Empty, s pluginapi.DevicePlugin_ListAndWatchServer) error {
-	log.Println("ListAndWatch")
+	glog.V(2).Infof("ListAndWatch")
```
Comment (vikaschoudhary16) [inline pkg/kubelet/cm/deviceplugin/device_plugin_stub.go]: same as below one.

File: pkg/kubelet/cm/deviceplugin/device_plugin_stub.go [hunk 3]
```diff
@@ -66,39 +86,57 @@ func (m *Stub) Start() error {
...
+				panic(err)
+			case <-m.stop:
+				m.server.GracefulStop()
+				return
+			}
 		}
-		time.Sleep(1 * time.Second)
+	}()
+
+	// Wait for server to start by launching a blocking connexion
```
Comment (vikaschoudhary16) [inline pkg/kubelet/cm/deviceplugin/device_plugin_stub.go]: typo: s/connexion/connection

**Review by vikaschoudhary16 — COMMENTED** [review]
Mostly lgtm. minor comment on log level used.

File: pkg/kubelet/cm/deviceplugin/device_plugin_stub.go (hunk 1 above)
Comment (RenaudWasTaken) [reply pkg/kubelet/cm/deviceplugin/device_plugin_stub.go]: I don't think log levels really matters for tests :)

File: pkg/kubelet/cm/deviceplugin/device_plugin_stub.go (hunk 3 above)
Comment (RenaudWasTaken) [reply pkg/kubelet/cm/deviceplugin/device_plugin_stub.go]: Thanks for noticing this!

**Comment by RenaudWasTaken** [comment]
FYI:
- [Current unit test time is 62.2s](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/52616/pull-kubernetes-bazel-test/19589/build-log.txt)
- [This PR paired with #52616 gets the unit test time to 2.2s](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/52612/pull-kubernetes-bazel-test/19402/build-log.txt)

That's a 28x time speedup

**Comment by tengqm** [comment]
@RenaudWasTaken this needs a rebase.

**Comment by RenaudWasTaken** [comment]
@feiskyer @pmorie @vikaschoudhary16 Rebased

**Pull request closed without merging**